### PR TITLE
Dice stats

### PR DESCRIPTION
**Pull request by L0neGamer**
adds a subcommand to roll that performs statistics on the given expression
the statistics are turned into a graph which is uploaded to discord

This is feature complete, allowing any expression to be turned into a distribution, which can then be analysed and prettified (ie turned into a graph and means and modal values extracted).

Other notable changes are additions to the docs, removal of the parameterised monad in funcinfobase, modifications to the expression parser to try and get better error messages (that failed, but keeping the changes), changes to the command parser to give better errors as well.

### CODE REVIEW

**Comment by L0neGamer** [comment]
https://github.com/WarwickTabletop/tablebot/issues/66 is the issue related to this branch and PR

File: stack.yaml [hunk 1]
```diff
@@ -51,6 +53,20 @@ extra-deps:
...
+- diagrams-lib-1.4.5.1
+- diagrams-postscript-1.5.1
+- diagrams-svg-1.4.3.1
+- svg-builder-0.1.1
+- active-0.2.0.15
+- dual-tree-0.2.3.0
+- monoid-extras-0.6.1
+- statestack-0.3
+- diagrams-rasterific-1.4.2.2
+- git: https://github.com/L0neGamer/haskell-distribution.git
```
Comment (L0neGamer) [inline stack.yaml:68]: This will need updating to just `distribution-1.1.1.1` once hackage has updated, but I don't know how long that will take

Comment (L0neGamer) [reply stack.yaml:69]: hackage has not updated, 20 days later. might be worth keeping it in then, or possibly pulling from the main repo? regardless, the updates within this commit are needed.

**Review by finnbar — CHANGES_REQUESTED** [review]
First off, this is all very cool and seems to generally hold up to testing - my comments are mainly documentation-related. I also have a few comments from my testing, which I reported to you on Discord but I'm repeating here:

* `!r 6d` reports two different errors, one with "couldn't" and another with "could not". Probably should be standardised, and maybe report one error if possible.
* Missing bracket errors (e.g. `!r maximum({d2, d2)` and `!r maximum({d2, d2}`) report the parent `WithError` rather than a more useful child error. Use `<??>` if you ever want to escape out of the `WithError`'s error. (Maybe comment on `<??>` with that?)
* `!r 2d(d6)` is incorrectly flagged as an error.

Otherwise, looks amazing!

**Review by finnbar — APPROVED** [review]
Approved with changes made.

**Pull request merged**